
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>src.embed module &#8212; pathogen-embed  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="static/styles/theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="static/styles/bootstrap.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
<link href="static/styles/pydata-sphinx-theme.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />

  
  <link href="static/vendor/fontawesome/6.5.1/css/all.min.css?digest=8d27b9dea8ad943066ae" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="static/vendor/fontawesome/6.5.1/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="static/vendor/fontawesome/6.5.1/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="static/vendor/fontawesome/6.5.1/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="static/styles/sphinx-book-theme.css?v=384b581d" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae" />
<link rel="preload" as="script" href="static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae" />
  <script src="static/vendor/fontawesome/6.5.1/js/all.min.js?digest=8d27b9dea8ad943066ae"></script>

    <script src="static/documentation_options.js?v=9eb32ce0"></script>
    <script src="static/doctools.js?v=888ff710"></script>
    <script src="static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="static/scripts/sphinx-book-theme.js?v=efea14e4"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'index';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a id="pst-skip-link" class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <header class="bd-header navbar navbar-expand-lg bd-navbar">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="#">
  
  
  
  
  
  
    <p class="title logo__title">pathogen-embed  documentation</p>
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="sources/index.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>src.embed module</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-line-interface">Command line interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pathogen-embed">pathogen-embed</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments">Positional Arguments</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments">Named Arguments</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Sub-commands">Sub-commands</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pca">pca</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat1">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#t-sne">t-sne</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat2">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#umap">umap</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat3">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mds">mds</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat4">Named Arguments</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pathogen-distance">pathogen-distance</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pathogen-cluster">pathogen-cluster</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-src.pathogen_embed">API</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="src-embed-module">
<h1>src.embed module<a class="headerlink" href="#src-embed-module" title="Link to this heading">#</a></h1>
<section id="command-line-interface">
<h2>Command line interface<a class="headerlink" href="#command-line-interface" title="Link to this heading">#</a></h2>
</section>
<section id="pathogen-embed">
<h2>pathogen-embed<a class="headerlink" href="#pathogen-embed" title="Link to this heading">#</a></h2>
<p><p>Reduced dimension embeddings for pathogen sequences</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">pathogen</span><span class="o">-</span><span class="n">embed</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">--</span><span class="n">alignment</span> <span class="n">ALIGNMENT</span> <span class="p">[</span><span class="n">ALIGNMENT</span> <span class="o">...</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">distance</span><span class="o">-</span><span class="n">matrix</span> <span class="n">DISTANCE_MATRIX</span> <span class="p">[</span><span class="n">DISTANCE_MATRIX</span> <span class="o">...</span><span class="p">]]</span> <span class="p">[</span><span class="o">--</span><span class="n">separator</span> <span class="n">SEPARATOR</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">indel</span><span class="o">-</span><span class="n">distance</span><span class="p">]</span>
                      <span class="p">[</span><span class="o">--</span><span class="n">random</span><span class="o">-</span><span class="n">seed</span> <span class="n">RANDOM_SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">dataframe</span> <span class="n">OUTPUT_DATAFRAME</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">figure</span> <span class="n">OUTPUT_FIGURE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">embedding</span><span class="o">-</span><span class="n">parameters</span> <span class="n">EMBEDDING_PARAMETERS</span><span class="p">]</span>
                      <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">pairwise</span><span class="o">-</span><span class="n">distance</span><span class="o">-</span><span class="n">figure</span> <span class="n">OUTPUT_PAIRWISE_DISTANCE_FIGURE</span><span class="p">]</span>
                      <span class="p">{</span><span class="n">pca</span><span class="p">,</span><span class="n">t</span><span class="o">-</span><span class="n">sne</span><span class="p">,</span><span class="n">umap</span><span class="p">,</span><span class="n">mds</span><span class="p">}</span> <span class="o">...</span>
</pre></div>
</div>
<section id="positional-arguments">
<h3>Positional Arguments<a class="headerlink" href="#positional-arguments" title="Link to this heading">#</a></h3>
<dl class="option-list">
<dt><kbd>command</kbd></dt>
<dd><p>Possible choices: pca, t-sne, umap, mds</p>
</dd>
</dl>
</section>
<section id="named-arguments">
<h3>Named Arguments<a class="headerlink" href="#named-arguments" title="Link to this heading">#</a></h3>
<dl class="option-list">
<dt><kbd>--alignment</kbd></dt>
<dd><p>an aligned FASTA file (or files) to create a distance matrix with. Make sure the strain order in this file matches the order in the distance matrix. If adding more than one alignment, make sure the order of the strains and strain names are the same between all the files.</p>
</dd>
<dt><kbd>--distance-matrix</kbd></dt>
<dd><p>a distance matrix (or matrices) that can be read in by pandas, index column as row 0.  If adding more than one distance matrix, make sure the order of the strains and strain names in the header are the same between all the files.</p>
</dd>
<dt><kbd>--separator</kbd></dt>
<dd><p>separator between columns in the given distance matrix</p>
<p>Default: “,”</p>
</dd>
<dt><kbd>--indel-distance</kbd></dt>
<dd><p>include insertions/deletions in genetic distance calculations</p>
<p>Default: False</p>
</dd>
<dt><kbd>--random-seed</kbd></dt>
<dd><p>an integer used for reproducible results.</p>
<p>Default: 314159</p>
</dd>
<dt><kbd>--output-dataframe</kbd></dt>
<dd><p>a csv file outputting the embedding with the strain name and its components.</p>
</dd>
<dt><kbd>--output-figure</kbd></dt>
<dd><p>outputs a plot of the embedding</p>
</dd>
<dt><kbd>--embedding-parameters</kbd></dt>
<dd><p>The file containing the parameters by which to tune the embedding. The values from the first record of this file will override default values or values provided by the command line arguments.</p>
</dd>
<dt><kbd>--output-pairwise-distance-figure</kbd></dt>
<dd><p>a scatterplot correlating the genetic vs Euclidean distances</p>
</dd>
</dl>
</section>
<section id="Sub-commands">
<h3>Sub-commands<a class="headerlink" href="#Sub-commands" title="Link to this heading">#</a></h3>
<section id="pca">
<h4>pca<a class="headerlink" href="#pca" title="Link to this heading">#</a></h4>
<p>Principal Component Analysis</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pathogen</span><span class="o">-</span><span class="n">embed</span> <span class="n">pca</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">components</span> <span class="n">COMPONENTS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">explained</span><span class="o">-</span><span class="n">variance</span> <span class="n">EXPLAINED_VARIANCE</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments_repeat1">
<h5>Named Arguments<a class="headerlink" href="#named-arguments_repeat1" title="Link to this heading">#</a></h5>
<dl class="option-list">
<dt><kbd>--components</kbd></dt>
<dd><p>the number of components for PCA</p>
<p>Default: 10</p>
</dd>
<dt><kbd>--explained-variance</kbd></dt>
<dd><p>the path for the CSV explained variance for each component</p>
</dd>
</dl>
</section>
</section>
<section id="t-sne">
<h4>t-sne<a class="headerlink" href="#t-sne" title="Link to this heading">#</a></h4>
<p>t-distributed Stochastic Neighborhood Embedding</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pathogen</span><span class="o">-</span><span class="n">embed</span> <span class="n">t</span><span class="o">-</span><span class="n">sne</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">components</span> <span class="n">COMPONENTS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">perplexity</span> <span class="n">PERPLEXITY</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">learning</span><span class="o">-</span><span class="n">rate</span> <span class="n">LEARNING_RATE</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments_repeat2">
<h5>Named Arguments<a class="headerlink" href="#named-arguments_repeat2" title="Link to this heading">#</a></h5>
<dl class="option-list">
<dt><kbd>--components</kbd></dt>
<dd><p>the number of components for t-SNE</p>
<p>Default: 2</p>
</dd>
<dt><kbd>--perplexity</kbd></dt>
<dd><p>The perplexity is related to the number of nearest neighbors. Because of this, the size of the dataset is proportional to the best perplexity value (large dataset -&gt; large perplexity). Values between 5 and 50 work best. The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: 30.0</p>
</dd>
<dt><kbd>--learning-rate</kbd></dt>
<dd><p>The learning rate for t-SNE is usually between 10.0 and 1000.0. Values out of these bounds may create innacurate results. The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: auto</p>
</dd>
</dl>
</section>
</section>
<section id="umap">
<h4>umap<a class="headerlink" href="#umap" title="Link to this heading">#</a></h4>
<p>Uniform Manifold Approximation and Projection</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pathogen</span><span class="o">-</span><span class="n">embed</span> <span class="n">umap</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">components</span> <span class="n">COMPONENTS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">nearest</span><span class="o">-</span><span class="n">neighbors</span> <span class="n">NEAREST_NEIGHBORS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">dist</span> <span class="n">MIN_DIST</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments_repeat3">
<h5>Named Arguments<a class="headerlink" href="#named-arguments_repeat3" title="Link to this heading">#</a></h5>
<dl class="option-list">
<dt><kbd>--components</kbd></dt>
<dd><p>the number of components for UMAP</p>
<p>Default: 2</p>
</dd>
<dt><kbd>--nearest-neighbors</kbd></dt>
<dd><p>Nearest neighbors controls how UMAP balances local versus global structure in the data (finer detail patterns versus global structure). This value is proportional to the size of the data (large dataset -&gt; large nearest neighbors). The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: 200</p>
</dd>
<dt><kbd>--min-dist</kbd></dt>
<dd><p>Minimum Distance controls how tightly packed the UMAP embedding is. While it does not change the structure of the data, it does change the embedding’s shape. The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: 0.5</p>
</dd>
</dl>
</section>
</section>
<section id="mds">
<h4>mds<a class="headerlink" href="#mds" title="Link to this heading">#</a></h4>
<p>Multidimensional Scaling</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pathogen</span><span class="o">-</span><span class="n">embed</span> <span class="n">mds</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">components</span> <span class="n">COMPONENTS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">stress</span> <span class="n">STRESS</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments_repeat4">
<h5>Named Arguments<a class="headerlink" href="#named-arguments_repeat4" title="Link to this heading">#</a></h5>
<dl class="option-list">
<dt><kbd>--components</kbd></dt>
<dd><p>the number of components for MDS</p>
<p>Default: 10</p>
</dd>
<dt><kbd>--stress</kbd></dt>
<dd><p>the path for the CSV stress for the embedding</p>
</dd>
</dl>
</section>
</section>
</section>
</section>
<section id="pathogen-distance">
<h2>pathogen-distance<a class="headerlink" href="#pathogen-distance" title="Link to this heading">#</a></h2>
<p><p>Hamming distance (optionally indel sensitive) similarity matrix for pathogen sequences</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">pathogen</span><span class="o">-</span><span class="n">distance</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">--</span><span class="n">alignment</span> <span class="n">ALIGNMENT</span> <span class="p">[</span><span class="o">--</span><span class="n">indel</span><span class="o">-</span><span class="n">distance</span><span class="p">]</span> <span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span>
</pre></div>
</div>
<section id="named-arguments">
<h3>Named Arguments<a class="headerlink" href="#named-arguments" title="Link to this heading">#</a></h3>
<dl class="option-list">
<dt><kbd>--alignment</kbd></dt>
<dd><p>an aligned FASTA file to create a distance matrix with. Make sure the strain order in this file matches the order in the distance matrix.</p>
</dd>
<dt><kbd>--indel-distance</kbd></dt>
<dd><p>include insertions/deletions in genetic distance calculations</p>
<p>Default: False</p>
</dd>
<dt><kbd>--output</kbd></dt>
<dd><p>a csv file outputting the distance matrix annotated with strain names as the columns</p>
</dd>
</dl>
</section>
</section>
<section id="pathogen-cluster">
<h2>pathogen-cluster<a class="headerlink" href="#pathogen-cluster" title="Link to this heading">#</a></h2>
<p><p>HDBSCAN clustering for reduced dimension embeddings</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">pathogen</span><span class="o">-</span><span class="n">cluster</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">--</span><span class="n">embedding</span> <span class="n">EMBEDDING</span> <span class="p">[</span><span class="o">--</span><span class="n">label</span><span class="o">-</span><span class="n">attribute</span> <span class="n">LABEL_ATTRIBUTE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">random</span><span class="o">-</span><span class="n">seed</span> <span class="n">RANDOM_SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">size</span> <span class="n">MIN_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">samples</span> <span class="n">MIN_SAMPLES</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="n">distance</span><span class="o">-</span><span class="n">threshold</span> <span class="n">DISTANCE_THRESHOLD</span><span class="p">]</span> <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">dataframe</span> <span class="n">OUTPUT_DATAFRAME</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">figure</span> <span class="n">OUTPUT_FIGURE</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments">
<h3>Named Arguments<a class="headerlink" href="#named-arguments" title="Link to this heading">#</a></h3>
<dl class="option-list">
<dt><kbd>--embedding</kbd></dt>
<dd><p>The embedding to assign clustering labels to via HDBSCAN (<a class="reference external" href="https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html">https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html</a>)</p>
</dd>
<dt><kbd>--label-attribute</kbd></dt>
<dd><p>the name of the cluster used to label the column in the resulting dataframe</p>
</dd>
<dt><kbd>--random-seed</kbd></dt>
<dd><p>an integer used for reproducible results.</p>
<p>Default: 314159</p>
</dd>
<dt><kbd>--min-size</kbd></dt>
<dd><p>minimum cluster size for HDBSCAN</p>
<p>Default: 5</p>
</dd>
<dt><kbd>--min-samples</kbd></dt>
<dd><p>minimum number of sample to seed a cluster for HDBSCAN. Lowering this value reduces number of samples that do not get clustered.</p>
<p>Default: 5</p>
</dd>
<dt><kbd>--distance-threshold</kbd></dt>
<dd><p>The float value for the distance threshold by which to cluster data in the embedding and assign labels via HDBSCAN. If no value is given in distance-threshold, the default distance threshold of 0.0 will be used.</p>
</dd>
<dt><kbd>--output-dataframe</kbd></dt>
<dd><p>a csv file outputting the embedding with the strain name and its components.</p>
</dd>
<dt><kbd>--output-figure</kbd></dt>
<dd><p>outputs a PDF with a plot of the embedding colored by cluster</p>
</dd>
</dl>
</section>
</section>
<section id="module-src.pathogen_embed">
<span id="api"></span><h2>API<a class="headerlink" href="#module-src.pathogen_embed" title="Link to this heading">#</a></h2>
<p>pathogen-embed.</p>
<p>Reduced dimension embeddings for pathogen sequences.</p>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-line-interface">Command line interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pathogen-embed">pathogen-embed</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#positional-arguments">Positional Arguments</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments">Named Arguments</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Sub-commands">Sub-commands</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pca">pca</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat1">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#t-sne">t-sne</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat2">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#umap">umap</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat3">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#mds">mds</a><ul class="nav section-nav flex-column">
<li class="toc-h5 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments_repeat4">Named Arguments</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pathogen-distance">pathogen-distance</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pathogen-cluster">pathogen-cluster</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#named-arguments">Named Arguments</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#module-src.pathogen_embed">API</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Sravani Nanduri and John Huddleston
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021–2024 Sravani Nanduri.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="static/scripts/bootstrap.js?digest=8d27b9dea8ad943066ae"></script>
<script src="static/scripts/pydata-sphinx-theme.js?digest=8d27b9dea8ad943066ae"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>