
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>src.embed module &#8212; pathogen-embed  documentation</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="static/pygments.css" />
    <link rel="stylesheet" href="static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="static/documentation_options.js"></script>
    <script src="static/jquery.js"></script>
    <script src="static/underscore.js"></script>
    <script src="static/doctools.js"></script>
    <script src="static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="#">
      
      
      
      <h1 class="site-logo" id="site-title">pathogen-embed  documentation</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search the docs ..." aria-label="Search the docs ..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Theme by the <a href="https://ebp.jupyterbook.org">Executable Book Project</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="sources/index.rst.txt"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.rst</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-line-interface">
   Command line interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pathogen-embed">
   pathogen-embed
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#positional-arguments">
     Positional Arguments
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#named-arguments">
     Named Arguments
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Sub-commands">
     Sub-commands
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#pca">
       pca
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#named-arguments_repeat1">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#t-sne">
       t-sne
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#named-arguments_repeat2">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#umap">
       umap
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#named-arguments_repeat3">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mds">
       mds
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#named-arguments_repeat4">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pathogen-distance">
   pathogen-distance
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#named-arguments">
     Named Arguments
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pathogen-cluster">
   pathogen-cluster
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#named-arguments">
     Named Arguments
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#module-src.pathogen_embed">
   API
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>src.embed module</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-line-interface">
   Command line interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pathogen-embed">
   pathogen-embed
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#positional-arguments">
     Positional Arguments
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#named-arguments">
     Named Arguments
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#Sub-commands">
     Sub-commands
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#pca">
       pca
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#named-arguments_repeat1">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#t-sne">
       t-sne
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#named-arguments_repeat2">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#umap">
       umap
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#named-arguments_repeat3">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
     <li class="toc-h4 nav-item toc-entry">
      <a class="reference internal nav-link" href="#mds">
       mds
      </a>
      <ul class="nav section-nav flex-column">
       <li class="toc-h5 nav-item toc-entry">
        <a class="reference internal nav-link" href="#named-arguments_repeat4">
         Named Arguments
        </a>
       </li>
      </ul>
     </li>
    </ul>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pathogen-distance">
   pathogen-distance
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#named-arguments">
     Named Arguments
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pathogen-cluster">
   pathogen-cluster
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#named-arguments">
     Named Arguments
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#module-src.pathogen_embed">
   API
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section id="src-embed-module">
<h1>src.embed module<a class="headerlink" href="#src-embed-module" title="Permalink to this headline">#</a></h1>
<section id="command-line-interface">
<h2>Command line interface<a class="headerlink" href="#command-line-interface" title="Permalink to this headline">#</a></h2>
</section>
<section id="pathogen-embed">
<h2>pathogen-embed<a class="headerlink" href="#pathogen-embed" title="Permalink to this headline">#</a></h2>
<p><p>Reduced dimension embeddings for pathogen sequences</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">pathogen</span><span class="o">-</span><span class="n">embed</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">--</span><span class="n">alignment</span> <span class="n">ALIGNMENT</span> <span class="p">[</span><span class="o">--</span><span class="n">distance</span><span class="o">-</span><span class="n">matrix</span> <span class="n">DISTANCE_MATRIX</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">separator</span> <span class="n">SEPARATOR</span><span class="p">]</span>
                      <span class="p">[</span><span class="o">--</span><span class="n">indel</span><span class="o">-</span><span class="n">distance</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">random</span><span class="o">-</span><span class="n">seed</span> <span class="n">RANDOM_SEED</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">dataframe</span> <span class="n">OUTPUT_DATAFRAME</span><span class="p">]</span>
                      <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">figure</span> <span class="n">OUTPUT_FIGURE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">embedding</span><span class="o">-</span><span class="n">parameters</span> <span class="n">EMBEDDING_PARAMETERS</span><span class="p">]</span>
                      <span class="p">{</span><span class="n">pca</span><span class="p">,</span><span class="n">t</span><span class="o">-</span><span class="n">sne</span><span class="p">,</span><span class="n">umap</span><span class="p">,</span><span class="n">mds</span><span class="p">}</span> <span class="o">...</span>
</pre></div>
</div>
<section id="positional-arguments">
<h3>Positional Arguments<a class="headerlink" href="#positional-arguments" title="Permalink to this headline">#</a></h3>
<dl class="option-list">
<dt><kbd>command</kbd></dt>
<dd><p>Possible choices: pca, t-sne, umap, mds</p>
</dd>
</dl>
</section>
<section id="named-arguments">
<h3>Named Arguments<a class="headerlink" href="#named-arguments" title="Permalink to this headline">#</a></h3>
<dl class="option-list">
<dt><kbd>--alignment</kbd></dt>
<dd><p>an aligned FASTA file to create a distance matrix with. Make sure the strain order in this file matches the order in the distance matrix.</p>
</dd>
<dt><kbd>--distance-matrix</kbd></dt>
<dd><p>a distance matrix that can be read in by pandas, index column as row 0</p>
</dd>
<dt><kbd>--separator</kbd></dt>
<dd><p>separator between columns in the given distance matrix</p>
<p>Default: “,”</p>
</dd>
<dt><kbd>--indel-distance</kbd></dt>
<dd><p>include insertions/deletions in genetic distance calculations</p>
<p>Default: False</p>
</dd>
<dt><kbd>--random-seed</kbd></dt>
<dd><p>an integer used for reproducible results.</p>
<p>Default: 314159</p>
</dd>
<dt><kbd>--output-dataframe</kbd></dt>
<dd><p>a csv file outputting the embedding with the strain name and its components.</p>
</dd>
<dt><kbd>--output-figure</kbd></dt>
<dd><p>outputs a PNG plot of the embedding</p>
</dd>
<dt><kbd>--embedding-parameters</kbd></dt>
<dd><p>The file containing the parameters by which to tune the embedding. The values from the first record of this file will override default values or values provided by the command line arguments.</p>
</dd>
</dl>
</section>
<section id="Sub-commands">
<h3>Sub-commands<a class="headerlink" href="#Sub-commands" title="Permalink to this headline">#</a></h3>
<section id="pca">
<h4>pca<a class="headerlink" href="#pca" title="Permalink to this headline">#</a></h4>
<p>Principal Component Analysis</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pathogen</span><span class="o">-</span><span class="n">embed</span> <span class="n">pca</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">components</span> <span class="n">COMPONENTS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">explained</span><span class="o">-</span><span class="n">variance</span> <span class="n">EXPLAINED_VARIANCE</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments_repeat1">
<h5>Named Arguments<a class="headerlink" href="#named-arguments_repeat1" title="Permalink to this headline">#</a></h5>
<dl class="option-list">
<dt><kbd>--components</kbd></dt>
<dd><p>the number of components for PCA</p>
<p>Default: 10</p>
</dd>
<dt><kbd>--explained-variance</kbd></dt>
<dd><p>the path for the CSV explained variance for each component</p>
</dd>
</dl>
</section>
</section>
<section id="t-sne">
<h4>t-sne<a class="headerlink" href="#t-sne" title="Permalink to this headline">#</a></h4>
<p>t-distributed Stochastic Neighborhood Embedding</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pathogen</span><span class="o">-</span><span class="n">embed</span> <span class="n">t</span><span class="o">-</span><span class="n">sne</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">components</span> <span class="n">COMPONENTS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">perplexity</span> <span class="n">PERPLEXITY</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">learning</span><span class="o">-</span><span class="n">rate</span> <span class="n">LEARNING_RATE</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments_repeat2">
<h5>Named Arguments<a class="headerlink" href="#named-arguments_repeat2" title="Permalink to this headline">#</a></h5>
<dl class="option-list">
<dt><kbd>--components</kbd></dt>
<dd><p>the number of components for t-SNE</p>
<p>Default: 2</p>
</dd>
<dt><kbd>--perplexity</kbd></dt>
<dd><p>The perplexity is related to the number of nearest neighbors. Because of this, the size of the dataset is proportional to the best perplexity value (large dataset -&gt; large perplexity). Values between 5 and 50 work best. The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: 30.0</p>
</dd>
<dt><kbd>--learning-rate</kbd></dt>
<dd><p>The learning rate for t-SNE is usually between 10.0 and 1000.0. Values out of these bounds may create innacurate results. The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: 200.0</p>
</dd>
</dl>
</section>
</section>
<section id="umap">
<h4>umap<a class="headerlink" href="#umap" title="Permalink to this headline">#</a></h4>
<p>Uniform Manifold Approximation and Projection</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pathogen</span><span class="o">-</span><span class="n">embed</span> <span class="n">umap</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">components</span> <span class="n">COMPONENTS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">nearest</span><span class="o">-</span><span class="n">neighbors</span> <span class="n">NEAREST_NEIGHBORS</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">dist</span> <span class="n">MIN_DIST</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments_repeat3">
<h5>Named Arguments<a class="headerlink" href="#named-arguments_repeat3" title="Permalink to this headline">#</a></h5>
<dl class="option-list">
<dt><kbd>--components</kbd></dt>
<dd><p>the number of components for UMAP</p>
<p>Default: 2</p>
</dd>
<dt><kbd>--nearest-neighbors</kbd></dt>
<dd><p>Nearest neighbors controls how UMAP balances local versus global structure in the data (finer detail patterns versus global structure). This value is proportional to the size of the data (large dataset -&gt; large nearest neighbors. The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: 200</p>
</dd>
<dt><kbd>--min-dist</kbd></dt>
<dd><p>Minimum Distance controls how tightly packed the UMAP embedding is. While it does not change the structure of the data, it does change the embedding’s shape. The default value is the value consistently the best for pathogen analyses, results from an exhaustive grid search.</p>
<p>Default: 0.5</p>
</dd>
</dl>
</section>
</section>
<section id="mds">
<h4>mds<a class="headerlink" href="#mds" title="Permalink to this headline">#</a></h4>
<p>Multidimensional Scaling</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pathogen</span><span class="o">-</span><span class="n">embed</span> <span class="n">mds</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">components</span> <span class="n">COMPONENTS</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments_repeat4">
<h5>Named Arguments<a class="headerlink" href="#named-arguments_repeat4" title="Permalink to this headline">#</a></h5>
<dl class="option-list">
<dt><kbd>--components</kbd></dt>
<dd><p>the number of components for MDS</p>
<p>Default: 10</p>
</dd>
</dl>
</section>
</section>
</section>
</section>
<section id="pathogen-distance">
<h2>pathogen-distance<a class="headerlink" href="#pathogen-distance" title="Permalink to this headline">#</a></h2>
<p><p>Hamming distance (optionally indel sensitive) similarity matrix for pathogen sequences</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">pathogen</span><span class="o">-</span><span class="n">distance</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">--</span><span class="n">alignment</span> <span class="n">ALIGNMENT</span> <span class="p">[</span><span class="o">--</span><span class="n">indel</span><span class="o">-</span><span class="n">distance</span><span class="p">]</span> <span class="o">--</span><span class="n">output</span> <span class="n">OUTPUT</span>
</pre></div>
</div>
<section id="named-arguments">
<h3>Named Arguments<a class="headerlink" href="#named-arguments" title="Permalink to this headline">#</a></h3>
<dl class="option-list">
<dt><kbd>--alignment</kbd></dt>
<dd><p>an aligned FASTA file to create a distance matrix with. Make sure the strain order in this file matches the order in the distance matrix.</p>
</dd>
<dt><kbd>--indel-distance</kbd></dt>
<dd><p>include insertions/deletions in genetic distance calculations</p>
<p>Default: False</p>
</dd>
<dt><kbd>--output</kbd></dt>
<dd><p>a csv file outputting the distance matrix annotated with strain names as the columns</p>
</dd>
</dl>
</section>
</section>
<section id="pathogen-cluster">
<h2>pathogen-cluster<a class="headerlink" href="#pathogen-cluster" title="Permalink to this headline">#</a></h2>
<p><p>HDBSCAN clustering for reduced dimension embeddings</p>
</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">usage</span><span class="p">:</span> <span class="n">pathogen</span><span class="o">-</span><span class="n">cluster</span> <span class="p">[</span><span class="o">-</span><span class="n">h</span><span class="p">]</span> <span class="o">--</span><span class="n">embedding</span> <span class="n">EMBEDDING</span> <span class="p">[</span><span class="o">--</span><span class="n">label</span><span class="o">-</span><span class="n">attribute</span> <span class="n">LABEL_ATTRIBUTE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">random</span><span class="o">-</span><span class="n">seed</span> <span class="n">RANDOM_SEED</span><span class="p">]</span>
                        <span class="p">[</span><span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">size</span> <span class="n">MIN_SIZE</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="nb">min</span><span class="o">-</span><span class="n">samples</span> <span class="n">MIN_SAMPLES</span><span class="p">]</span> <span class="p">[</span><span class="o">--</span><span class="n">distance</span><span class="o">-</span><span class="n">threshold</span> <span class="n">DISTANCE_THRESHOLD</span><span class="p">]</span>
                        <span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">dataframe</span> <span class="n">OUTPUT_DATAFRAME</span> <span class="p">[</span><span class="o">--</span><span class="n">output</span><span class="o">-</span><span class="n">figure</span> <span class="n">OUTPUT_FIGURE</span><span class="p">]</span>
</pre></div>
</div>
<section id="named-arguments">
<h3>Named Arguments<a class="headerlink" href="#named-arguments" title="Permalink to this headline">#</a></h3>
<dl class="option-list">
<dt><kbd>--embedding</kbd></dt>
<dd><p>The embedding to assign clustering labels to via HDBSCAN (<a class="reference external" href="https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html">https://hdbscan.readthedocs.io/en/latest/how_hdbscan_works.html</a>)</p>
</dd>
<dt><kbd>--label-attribute</kbd></dt>
<dd><p>the name of the cluster used to label the column in the resulting dataframe</p>
</dd>
<dt><kbd>--random-seed</kbd></dt>
<dd><p>an integer used for reproducible results.</p>
<p>Default: 314159</p>
</dd>
<dt><kbd>--min-size</kbd></dt>
<dd><p>minimum cluster size for HDBSCAN</p>
<p>Default: 5</p>
</dd>
<dt><kbd>--min-samples</kbd></dt>
<dd><p>minimum number of sample to seed a cluster for HDBSCAN. Lowering this value reduces number of samples that do not get clustered.</p>
<p>Default: 5</p>
</dd>
<dt><kbd>--distance-threshold</kbd></dt>
<dd><p>The float value for the distance threshold by which to cluster data in the embedding and assign labels via HDBSCAN. If no value is given in distance-threshold, the default distance threshold of 0.0 will be used.</p>
</dd>
<dt><kbd>--output-dataframe</kbd></dt>
<dd><p>a csv file outputting the embedding with the strain name and its components.</p>
</dd>
<dt><kbd>--output-figure</kbd></dt>
<dd><p>outputs a PDF with a plot of the embedding colored by cluster</p>
</dd>
</dl>
</section>
</section>
<section id="module-src.pathogen_embed">
<span id="api"></span><h2>API<a class="headerlink" href="#module-src.pathogen_embed" title="Permalink to this headline">#</a></h2>
<p>pathogen-embed.</p>
<p>Reduced dimension embeddings for pathogen sequences.</p>
</section>
</section>


              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Sravani Nanduri and John Huddleston<br/>
  
      &copy; Copyright 2021–2023 Sravani Nanduri.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>